compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -66,9 +66,44 @@ object Objects:
 
   // ----------------------------- abstract domain -----------------------------
 
+  /** Syntax for the data structure abstraction used in abstract domain:
+   *
+   * ve ::= ObjectRef(class)
+   *      | OfClass(class, vs[outer], ctor, args, env)
+   *      | OfArray(object[owner], regions)
+   *      | Fun(..., env)
+   *      | Cold                                                         // abstract values in domain
+   * vs ::= Set(ve)                                                      // set of abstract values
+   *
+   * refMap = ( ObjectRef | OfClass ) -> ( valsMap, varsMap, outersMap ) // refMap stores field informations of an object or instance
+   * valsMap = valsym -> vs                                              // maps immutable fields to their values
+   * varsMap = valsym -> addr                                            // each mutable field has an abstract address
+   * outersMap = class -> vs                                             // maps outer objects to their values
+   *
+   * arrayMap = OfArray -> addr                                          // an array has one address that stores the join value of every element
+   *
+   * heap = addr -> vs                                                   // only this map's entries can be modified
+   *
+   * env = (valsMap, Option[env])                                        // stores local variables in the residing method, and possibly outer environments
+   *
+   * addr ::= localVarAddr(regions, valsym, owner)
+   *        | fieldVarAddr(regions, valsym, owner)                       // independent of OfClass/ObjectRef
+   *        | arrayAddr(regions, owner)                                  // independent of array element type
+   *
+   * regions ::= List(sourcePosition)
+   */
+
   sealed abstract class Value:
     def show(using Context): String
+    def widen(height : Int)(using Context): Value =
+      if height == 0 then Cold else this
+
+  /** ValueElement are elements that can be contained in a RefSet */
+  sealed abstract class ValueElement extends Value
 
+  sealed trait ThisValue extends Value:
+    override def widen(height : Int)(using Context) : ThisValue =
+      if height == 0 then Cold else this
```

Comment:
I'd suggest merging `widen` code into as fewer methods as possible as before, so that it's easier to reason about the pre/pos conditions.

Reply:
We can discuss with Ondřej for this later, as he suggests to rewrite `widen` as an abstract method for `Value`